src/main/java/org/spongepowered/common/registry/type/MutableCatalogRegistryModule.java

```diff
@@ -0,0 +1,142 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.spongepowered.common.registry.type;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import org.spongepowered.api.CatalogType;
+import org.spongepowered.api.registry.AlternateCatalogRegistryModule;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+public class MutableCatalogRegistryModule<T extends CatalogType> implements AlternateCatalogRegistryModule<T> {
+
+    private final Map<String, T> mapping;
+    private final Function<String, String> keyTransformer;
+
+    public MutableCatalogRegistryModule() {
+        this(Maps.newHashMap());
+    }
+
+    public MutableCatalogRegistryModule(Map<String, T> mapping) {
+        this.mapping = checkNotNull(mapping, "mapping");
+        this.keyTransformer = (input) -> input;
+    }
+
+    public MutableCatalogRegistryModule(Function<String, String> keyTransformer) {
+        this(Maps.newHashMap(), keyTransformer);
+    }
+
+    public MutableCatalogRegistryModule(Map<String, T> mapping, Function<String, String> keyTransformer) {
+        this.mapping = checkNotNull(mapping, "mapping");
+        this.keyTransformer = checkNotNull(keyTransformer, "keyTransformer");
+    }
+
+    @Override
+    public final Optional<T> getById(String id) {
+        return Optional.ofNullable(this.mapping.get(mapId(checkNotNull(id, "id"))));
+    }
+
+    private String mapId(String id) {
+        return this.keyTransformer.apply(id.toLowerCase());
+    }
+
+    @Override
+    public final Collection<T> getAll() {
+        return ImmutableSet.copyOf(this.mapping.values());
+    }
+
+    @Override
+    public Map<String, T> provideCatalogMap() {
+        return ImmutableMap.copyOf(this.mapping);
+    }
+
+    @SuppressWarnings("unchecked")
+    protected final void registerUnsafe(Object type, String... aliases) {
+        register((T) type, aliases);
+    }
+
+    protected void register(T type, String... aliases) {
+        checkNotNull(type, "type");
+        checkNotNull(aliases, "aliases");
+        final String id = checkNotNull(type.getId(), "id");
+        checkArgument(!id.isEmpty(), "Passed type's id is empty!");
+        this.mapping.put(id.toLowerCase(), type);
```

Comment:
Oh nevermind this isn't replacing the AdditionalRegistryModule so isn't available for plugin registry anyway.